### PR TITLE
Bug 1360121 - Youtube video will not appear in history list

### PR DESCRIPTION
**Pull request by justindarc**

### CODE REVIEW

**Comment by farhanpatel** [comment]
### SwiftLint found issues

#### Warnings

File | Line | Reason |
| --- | ----- | ----- |
BrowserViewController.swift | 932 | Force casts should be avoided. 
BrowserViewController.swift | 1930 | Force casts should be avoided. 
BrowserViewController.swift | 1934 | Force casts should be avoided. 
BrowserViewController.swift | 1891 | `where` clauses are preferred over a single `if` inside a `for`. 
BrowserViewController.swift | 1067 | Prefer `_ = foo()` over `let _ = foo()` when discarding a result from a function. 
BrowserViewController.swift | 1084 | Prefer `_ = foo()` over `let _ = foo()` when discarding a result from a function. 
BrowserViewController.swift | 2904 | Prefer `_ = foo()` over `let _ = foo()` when discarding a result from a function. 


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by farhanpatel** [comment]
There is a `NavigationTest.swift` mentioned that doesnt seem to be in this PR. 
Also people can use the pushState to forge history with this I think. Look at the comment here 
https://github.com/mozilla-mobile/firefox-ios/blob/master/Client/Frontend/Browser/BrowserViewController.swift#L953
I think the same comment would apply here no?

**Comment by justindarc** [comment]
@farhanpatel good catch. I don't know how the project file got all screwed up, but I'll fix it.

Also, let me write a test case to double check that this doesn't create any issues with `pushState`. I don't *think* it would because I don't believe you can't change origins with `pushState`, but we should definitely check it to be certain.

File: Client/Assets/HistoryStateHelper.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+(function() {
+    var nativeHistoryPushState = window.history.pushState;
+    var nativeHistoryReplaceState = window.history.replaceState;
+
+    window.history.pushState = function(state, title, url) {
```
Comment (farhanpatel) [inline Client/Assets/HistoryStateHelper.js]: A comment why this exists would be helpful for the future.

File: Client/Frontend/Browser/HistoryStateHelper.swift [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+import Foundation
+import WebKit
+
+protocol HistoryStateHelperDelegate: class {
+    func historyStateHelper(_ historyStateHelper: HistoryStateHelper, didPushOrReplaceStateInTab tab: Tab)
+}
+
+class HistoryStateHelper: TabHelper {
+    weak var delegate: HistoryStateHelperDelegate?
```
Comment (farhanpatel) [inline Client/Frontend/Browser/HistoryStateHelper.swift]: A comment here would also be helpful.

**Pull request merged**